### ISSUE TITLE
New trap behavior broke incomplete downloads

### ISSUE THREAD

**Issue by konimex** [kiss-community/kiss#25]
When downloading a file is interrupted, the offending file normally is removed. However, since certain versions, it no longer removes the incomplete file when the interruption happens, resulting in checksum mismatch.

After bisecting, I found https://github.com/kiss-community/kiss/commit/bdd95ba33ad208693ab4174d7d472d6fc003ebc4 to break the old behavior.

**Comment by dilyn-corner** [comment]
Indeed indeed... 
The difference is...
{old}
```
-> llvm Downloading sources
+ mkdir -p /home/dilyn/.cache/kiss/sources/llvm
+ cd /home/dilyn/.cache/kiss/sources/llvm
+ read -r src dest
+ [ -z https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz ]
+ [ -z https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz ]
+ [ -f ././llvmorg-12.0.0.tar.gz ]
+ [ -z  ]
+ log llvm Downloading https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz
+ printf %b%s %b%s%b %s\n \033[1;33m -> \033[m\033[1;34m llvm \033[m Downloading https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz
-> llvm Downloading https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz
+ mkdir -p /home/dilyn/.cache/kiss/sources/llvm/
+ curl https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz -fLo ././llvmorg-12.0.0.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   133  100   133    0     0    527      0 --:--:-- --:--:-- --:--:--   527
100 3542k    0 3542k    0     0  2426k      0 --:--:--  0:00:01 --:--:-- 3556k^C+ pkg_clean
+ [  = 1 ]
+ rm -rf /tmp/16979
+ exit 1
+ pkg_clean
+ [  = 1 ]
+ rm -rf /tmp/16979
```

{new}
```
-> llvm Downloading sources
+ mkdir -p /home/dilyn/.cache/kiss/sources/llvm
+ cd /home/dilyn/.cache/kiss/sources/llvm
+ read -r src dest
+ [ -z https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz ]
+ [ -z https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz ]
+ [ -f ././llvmorg-12.0.0.tar.gz ]
+ [ -z  ]
+ log llvm Downloading https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz
+ printf %b%s %b%s%b %s\n \033[1;33m -> \033[m\033[1;34m llvm \033[m Downloading https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz
-> llvm Downloading https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz
+ mkdir -p /home/dilyn/.cache/kiss/sources/llvm/
+ curl https://github.com/llvm/llvm-project/archive/llvmorg-12.0.0.tar.gz -fLo ././llvmorg-12.0.0.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   133  100   133    0     0    527      0 --:--:-- --:--:-- --:--:--   527
100 3542k    0 3542k    0     0  2426k      0 --:--:--  0:00:01 --:--:-- 3556k^C+ pkg_clean
+ [  = 1 ]
+ rm -rf /tmp/16979
+ exit 1
+ pkg_clean
+ [  = 1 ]
+ rm -rf /tmp/16979
```

**Comment by cemkeylan** [comment]
Yes, I have also noticed the error recently. Most of the time you want
the package manager to cleanup and exit immediately, and sometimes you
don't. I made this change initially because the package manager would
ignore interrupt calls when someone interrupted during update
enumeration. The best solution for this would be adding a separate trap
state, for places where you want to handle the interrupts on your own.

What I am thinking of is adding a small function for managing trap
states. This way they will be easier to control. I'll send a patch when
I can.

--
Best Regards,
Cem Keylan

**Comment by aaronNGi** [comment]
> Most of the time you want the package manager to cleanup and exit immediately, and sometimes you don't.

I was wondering about reasons why we don't want to exit immediately but haven't dug deep enough yet. Would be nice if you could elaborate.

**Comment by dilyn-corner** [comment]
This has been fixed edfb25aa2da44076dcb35b19f8e6cfddd5a66dfa Thanks Cem :)
